File: 백준/Gold/9370. 미확인 도착지/미확인 도착지.java
import java.util.*;
import java.io.*;

class Point implements Comparable<Point>{
    int dis;
    int idx;
    public Point(int dis, int idx) {
        this.dis = dis;
        this.idx = idx;
    }

    @Override
    public int compareTo(Point other) {
        return Integer.compare(this.dis, other.dis);
    }

}

public class Main {
    static int T;
    static int n, m, t, s, g, h, a, b, d;
    static ArrayList<ArrayList<Point>> graph;
    static int[] first, gDijk, hDijk;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        T = Integer.parseInt(br.readLine());
        for (int tc = 0; tc < T; tc++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            n = Integer.parseInt(st.nextToken());
            m = Integer.parseInt(st.nextToken());
            t = Integer.parseInt(st.nextToken());
            st = new StringTokenizer(br.readLine());
            s = Integer.parseInt(st.nextToken());
            g = Integer.parseInt(st.nextToken());
            h = Integer.parseInt(st.nextToken());

            graph = new ArrayList<>();
            for (int i = 0; i < n + 1; i++) {
                ArrayList<Point> arr = new ArrayList<>();
                graph.add(arr);
            }

            for (int i = 0; i < m; i++) {
                st = new StringTokenizer(br.readLine());
                a = Integer.parseInt(st.nextToken());
                b = Integer.parseInt(st.nextToken());
                d = Integer.parseInt(st.nextToken());
                graph.get(a).add(new Point(d, b));
                graph.get(b).add(new Point(d, a));
            }

            int[] target = new int[t];
            for (int i = 0; i < t; i++) {
                int tg = Integer.parseInt(br.readLine());
                target[i] = tg;
            }
            first = dijkstra(s);

            gDijk = dijkstra(g);
            hDijk = dijkstra(h);

            ArrayList<Integer> answer = new ArrayList<>();
            for (int tg : target) {
                if (first[g] + gDijk[h] + hDijk[tg] == first[tg] || first[h] + hDijk[g] + gDijk[tg] == first[tg]) {
                    answer.add(tg);
                }
            }

            Collections.sort(answer);
            for (int ans : answer) {
                sb.append(ans + " ");
            }
            sb.append("\n");

        }
        System.out.println(sb);
    }

    static int[] dijkstra(int start) {
        PriorityQueue<Point> q = new PriorityQueue<>();
        int[] distance = new int[n+1];
        Arrays.fill(distance, Integer.MAX_VALUE);
        q.offer(new Point(0, start));
        distance[start] = 0;
        while (!q.isEmpty()) {
            Point now = q.poll();
            if (now.dis > distance[now.idx]) {
                continue;
            }
            for (Point next : graph.get(now.idx)) {
                int cost = now.dis + next.dis;
                if (cost < distance[next.idx]) {
                    q.add(new Point(cost, next.idx));
                    distance[next.idx] = cost;
                }
            }
        }
        return distance;
    }
}

File: 백준/Gold/9370. 미확인 도착지/미확인 도착지.py
from heapq import heappush, heappop
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**9)

def dijkstra(start):
    q = []
    # 최단거리 테이블을 모두 무한으로 초기화
    distance = [INF] * (n+1)
    # 시작노드로 가기 위한 최단 경로는 0으로 설정하고 큐에 삽입
    distance[start] = 0
    heappush(q, (0, start))

    while q:
        # 가장 거리가 짧은 노드 가져오기
        dis, node = heappop(q)
        # 현재 노드가 처리된 노드라면 패스
        if distance[node] < dis:
            continue

        # 현재 노드와 연결된 다른 인접한 노드 확인
        for next_node in graph[node]:
            cost = dis + next_node[1]
            # 현재 노드를 거쳐서 가는게 거리가 더 짧을 경우
            if cost < distance[next_node[0]]:
                distance[next_node[0]] = cost
                heappush(q, (cost, next_node[0]))
    return distance

    pass

T = int(input())
for _ in range(T):
    n, m, t = map(int, input().split()) # 교차로 수, 도로 수, 목적지 후보 개수
    s, g, h = map(int, input().split()) # 출발지, g, h
    graph = [[] for _ in range(n+1)]
    INF = int(1e9)
    for _ in range(m):
        a, b, d = map(int, input().split())
        graph[a].append((b, d))
        graph[b].append((a, d))
    target = [int(input()) for _ in range(t)]

    # s랑 g 사이의 최소거리
    # g랑 target들 사이의 최소거리
    # s랑 h 사이의 최소거리
    # h랑 target들 사이의 최소거리

    # 다익스트라
    first = dijkstra(s)

    g_dijkstra = dijkstra(g)
    h_dijkstra = dijkstra(h)

    arr = []
    for tg in target:
        if first[g] + g_dijkstra[h] + h_dijkstra[tg] == first[tg] or first[h] + h_dijkstra[g] + g_dijkstra[tg] == first[tg]:
            arr.append(tg)

    arr.sort()
    print(*arr)
